File: Irrgarten/src/irrgarten/Labyrinth.java
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author Sergio Salvador Gil
 */
public class Labyrinth {
    
        
    private final char BLOCK_CHAR='X';
    private final char EMPTY_CHAR='-';
    private final char MONSTER_CHAR='M';
    private final char COMBAT_CHAR='C';
    private final char EXIT_CHAR='E';
    private final int ROW=0;
    private final int COL=1;
    
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;
    
    private Monster[][] monsters;
    private Player[][] players;
    private char[][]labyrinth;
    
    Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows=nRows;
        this.nCols=nCols;
        this.exitRow=exitRow;
        this.exitCol=exitCol;
        
        monsters = new Monster[nRows][nCols];
        players = new Player[nRows][nCols];
        labyrinth = new char[nRows][nCols];
        
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                labyrinth[i][j] = EMPTY_CHAR;
            }
        }
        
        this.labyrinth[exitRow][exitCol] = EXIT_CHAR;
    }
    
    public void spreadPlayers(ArrayList<Player> players){
        
        for (int i = 0; i < players.size(); i++){
            Player p = players.get(i);
            int [] pos = this.randomEmptyPos();
            putPlayer2D(-1,-1,pos[ROW],pos[COL],p);
        }
    }
    
    public boolean haveAWinner(){
        return players[exitRow][exitCol] != null;
    }
    
    public String toString(){
        String laberinto = "";
        for (int i = 0; i< nCols; i++){
            for (int j = 0; j < nRows; j++){
                if (posOK(i,j))
                    laberinto += "[" + labyrinth[i][j] + "]";
            }
            laberinto += "\n";
        }
        return laberinto;
    }
    
    public void addMonster(int row, int col, Monster monster){
        if (posOK(row, col)){
            labyrinth[row][col]=MONSTER_CHAR;
            monsters[row][col]=monster;
            monster.setPos(row,col);
        }
    }
    
    public Monster putPlayer(Directions direction, Player player){
        int oldRow=player.getRow();
        int oldCol=player.getCol();
        
        int[]newPos=this.dir2Pos(oldRow,oldCol,direction);
        Monster monster = this.putPlayer2D(oldRow,oldCol,newPos[ROW],newPos[COL],player);
        return monster;
    }
    
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow, incCol;
        if (orientation == Orientation.VERTICAL){
            incRow=1;
            incCol=0;
        }
        else{
            incRow=0;
            incCol=1;
        }
        int row=startRow;
        int col=startCol;
        
        while(posOK(row,col) && emptyPos(row,col) && length>0){
            labyrinth[row][col]=BLOCK_CHAR;
            
            length-=1;
            row+=incRow;
            col+=incCol;
        }
    }
    
    public ArrayList<Directions> validMoves(int row, int col){
        ArrayList<Directions> output = new ArrayList<>();
        
        if (canStepOn(row+1,col)){
            output.add(Directions.DOWN);
        }
        
        if (canStepOn(row-1,col)){
            output.add(Directions.UP);
        }
        
        if (canStepOn(row,col+1)){
            output.add(Directions.RIGHT);
        }
        
        if (canStepOn(row,col+1)){
            output.add(Directions.LEFT);
        }
        return output;
    }
    
    // Métodos privados
    
    private boolean posOK(int row, int col){
        return (0 <= row && row < nRows && 0 <= col && col < nCols);
    }
    
    private boolean emptyPos(int row, int col){
        return labyrinth[row][col]==EMPTY_CHAR;
    }
    
    private boolean monsterPos(int row, int col){
        return labyrinth[row][col]==MONSTER_CHAR;
    }
    
    private boolean exitPos(int row, int col){
        return labyrinth[row][col]==EXIT_CHAR;
    }
    
    private boolean combatPos(int row, int col){
        return labyrinth[row][col]==COMBAT_CHAR;
    }
    
    private boolean canStepOn(int row, int col){
        return (posOK(row,col)&&(monsterPos(row, col)||exitPos(row,col)||combatPos(row,col)));
    }
    
    private void updateOldPos(int row, int col){
        if (posOK(row,col)){
            if (combatPos(row,col)){
                labyrinth[row][col]=MONSTER_CHAR;
            }
            else labyrinth[row][col]=EMPTY_CHAR;
        }
    }
    
    private int [] dir2Pos(int row, int col, Directions direction){
        
        int pos[]={0,0};
        if(direction==Directions.DOWN){
           pos[ROW]=row+1;
           pos[COL]=col;
        }
        if(direction==Directions.LEFT){
           pos[ROW]=row;
           pos[COL]=col-1;
        }
        if(direction==Directions.RIGHT){
           pos[ROW]=row;
           pos[COL]=col+1;
        }
        if(direction==Directions.UP){
           pos[ROW]=row-1;
           pos[COL]=col;
        }
        return pos;
    }
    
    private int[] randomEmptyPos(){
        boolean empty=false;
        Dice dado = new Dice();
        int pos[]={0,0};
        
        while (empty=false){
            pos[ROW]=dado.randomPos(nRows);
            pos[COL]=dado.randomPos(nCols);
            empty=emptyPos(pos[ROW],pos[COL]);
        }
        return pos;
    }

    
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
       Monster output=null;
       
       if (canStepOn(row,col)){
           if (posOK(row,col)){
               Player p = players[oldRow][oldCol];
               if (p==player){
                   this.updateOldPos(oldRow, oldCol);
                   players[oldRow][oldCol]= null;
               }
           }
           
           boolean monsterPos = this.monsterPos(row,col);
           
           if (monsterPos){
               labyrinth[row][col]=COMBAT_CHAR;
               output = monsters[row][col];
           }
           else{
               char number=player.getNumber();
               labyrinth[row][col]=number;
           }
           
           players[row][col]=player;
           player.setPos(row,col);
       }
       return output;
    }
}

File: rb/Irrgarten/Labyrinth.rb
#encoding:UTF-8

module Irrgarten
    class Labyrinth
        @@BLOCK_CHAR='X'
        @@EMPTY_CHAR='-'
        @@MONSTER_CHAR='M'
        @@COMBAT_CHAR='C'
        @@EXIT_CHAR='E'
        @@ROW=0
        @@COL=1

        def initialize(n_rows, n_cols, exit_row, exit_col)
            @n_rows=n_rows
            @n_cols=n_cols
            @exit_row=exit_row
            @exit_col=exit_col

            @monsters = Array.new(n_rows) { Array.new(n_cols) }
            @players = Array.new(n_rows) { Array.new(n_cols) }
            @labyrinth = Array.new(n_rows) { Array.new(n_cols, '-') }

            @labyrinth[@exit_row][@exit_col]='E'
        end

        def spread_players(players)
            players.each do |p|
                pos = random_empty_pos
                put_player_2d(-1,-1,pos[@@ROW],pos[@@COL],p)
            end
        end

        def have_a_winner
            return @players[@exit_row][@exit_col]!=nil
        end

        def to_s

            s=""

            @n_rows.times do |i|
                @n_cols.times do |j|
                    s+= "[#{@labyrinth[i][j]}]"
                end
                s+="\n"
            end
            return s
        end

        def add_monster(row,col,monster)
            if(pos_ok(row,col))
                @labyrinth[row][col]=@@MONSTER_CHAR
                @monsters[row][col]=monster
                monster.set_pos(row,col)
            end
        end

        def put_player(direction,player)
            old_row=player.get_row
            old_col=player.get_col

            new_pos=dir_2_pos(old_row,old_col,direction)
            put_player_2d(old_row,old_col,new_pos[@@ROW],new_pos[@@COL],player)
        end

        def add_block(orientation, start_row, start_col, length)
            if (orientation==Orientation::VERTICAL)
                inc_row=1
                inc_col=0
            else
                inc_row=0
                inc_col=1
            end
            row=start_row
            col=start_col
            while (pos_ok(row,col) && empty_pos(row,col) && length>0) do
                @labyrinth [row][col]=@@BLOCK_CHAR
                length=length-1
                row+=inc_row
                col+=inc_col
            end
        end

        def valid_moves(row,col)
            output = Array.new
            if (can_step_on(row+1,col))
                output << Directions::DOWN
            end
            if (can_step_on(row-1,col))
                output << Directions::UP
            end

            if (can_step_on(row,col+1))
                output << Directions::RIGHT
            end
            if (can_step_on(row,col-1))
                output << Directions::LEFT
            end
            output
        end


        private


        def pos_ok(row,col)
            return (0 <= row && row < @n_rows && 0 <= col && col < @n_cols)
        end

        def empty_pos(row,col)
            return @labyrinth[row][col]==@@EMPTY_CHAR;
        end

        def monster_pos(row,col)
            return @labyrinth[row][col]==@@MONSTER_CHAR;
        end

        def exit_pos(row,col)
            return @labyrinth[row][col]==@@EXIT_CHAR;
        end

        def combat_pos(row,col)
            return @labyrinth[row][col]==@@COMBAT_CHAR;
        end

        def can_step_on(row,col)
            return (pos_ok(row,col) && (empty_pos(row, col) || monster_pos(row, col) || exit_pos(row,col) || combat_pos(row,col)))
        end

        def update_old_pos(row,col)
            if (pos_ok(row,col))
                if(combat_pos(row,col))
                    @labyrinth[row][col]=@@MONSTER_CHAR
                else
                    @labyrinth[row][col]=@@EMPTY_CHAR
                end
            end
        end

        def dir_2_pos(row,col,direction)

            case direction
            when Directions::LEFT
                return [row, col - 1]
            when Directions::RIGHT
                return [row, col + 1]
            when Directions::UP
                return [row - 1, col]
            when Directions::DOWN
                return [row + 1, col]
            end
        end

        def random_empty_pos
            empty=false
            pos=[0,0]

            while (empty==false) do
                pos[@@ROW]=Dice.random_pos(@n_rows)
                pos[@@COL]=Dice.random_pos(@n_cols)
                empty=empty_pos(pos[@@ROW],pos[@@COL])
            end
            return pos
        end

        def put_player_2d(old_row,old_col,row,col,player)
            output=nil
            if (can_step_on(row,col))
                if (pos_ok(old_row,old_col))
                    p=@players[old_row][old_col]
                    if (p==player)
                        update_old_pos(old_row,old_col)
                        @players[old_row][old_col]=nil
                    end
                end
                monster_pos = monster_pos(row,col)
                if (monster_pos)
                    @labyrinth[row][col]=@@COMBAT_CHAR
                    output = @monsters[row][col]
                else
                    number=player.get_number
                    @labyrinth[row][col]=number
                end
                @players[row][col] = player
                player.set_pos(row, col)
            end
            output
        end

    end
end